src/modules/roc_netio/target_uv/roc_netio/transceiver.cpp

```diff
@@ -214,76 +206,80 @@ void Transceiver::stop_sem_cb_(uv_async_t* handle) {
     roc_panic_if_not(handle);
 
     Transceiver& self = *(Transceiver*)handle->data;
-    self.stop_();
-    self.close_();
+    self.stop_all_elements_();
+    self.close_sems_();
     self.process_tasks_();
 }
 
+void Transceiver::close_cb_(uv_handle_t* handle) {
+    roc_panic_if_not(handle);
+
+    Transceiver& self = *(Transceiver*)handle->data;
+
+    core::Mutex::Lock lock(self.mutex_);
+
+    if (handle == (uv_handle_t*)&self.stop_sem_) {
+        self.stop_sem_initialized_ = false;
+    } else {
+        self.task_sem_initialized_ = false;
+    }
+
+    if (self.stop_sem_initialized_ || self.task_sem_initialized_) {
+        return;
+    }
+
+    self.cond_.broadcast();
+}
+
 void Transceiver::stop_() {
-    for (core::SharedPtr<UDPReceiver> rp = receivers_.front(); rp;
-         rp = receivers_.nextof(*rp)) {
-        rp->stop();
+    core::Mutex::Lock lock(mutex_);
+
+    if (stopped_) {
+        return;
     }
 
-    for (core::SharedPtr<UDPSender> sp = senders_.front(); sp;
-         sp = senders_.nextof(*sp)) {
-        sp->stop();
+    if (int err = uv_async_send(&stop_sem_)) {
+        roc_panic("transceiver: uv_async_send(): [%s] %s", uv_err_name(err),
+                  uv_strerror(err));
     }
+
+    wait_stopped_();
+    stopped_ = true;
 }
 
-void Transceiver::close_() {
-    if (task_sem_initialized_) {
-        uv_close((uv_handle_t*)&task_sem_, NULL);
-        task_sem_initialized_ = false;
+void Transceiver::wait_stopped_() {
+    while (elements_.size() || task_sem_initialized_ || stop_sem_initialized_) {
+        cond_.wait();
     }
+}
 
-    if (stop_sem_initialized_) {
-        uv_close((uv_handle_t*)&stop_sem_, NULL);
-        stop_sem_initialized_ = false;
+void Transceiver::stop_all_elements_() {
+    for (core::SharedPtr<IElement> pp = elements_.front(); pp;
+         pp = elements_.nextof(*pp)) {
+        pp->stop();
```

Comment:
I suggest to rework this method and add a guarantee that it will work even if `pp->stop()` will invoke `Transceiver::close()`. We can do it like this:

```
void Transceiver::stop_all_elements_() {
    core::SharedPtr<IElement> curr = elements_.front();
    while (curr) {
        core::SharedPtr<IElement> next = elements_.nextof(*curr);
        curr->stop();
        curr = next;
    }
}
```

Reply:
How did you notice it :) Fixed

Reply:
It's a standard idiom: deleting an element inside a loop iterating the container.